Support configuring version per profile

RobinBoers opened the pull request:
In our project, we had a place where we were running multiple builds through multiple tailwind profiles, and one had already been updated to be compatible with v4 (and thus incompatible with v3), while the other is still stuck on v3 for now.

I added support to run a different version of the CLI for a profile. If no version is explicitly configured for the profile, it will fall back to the globally configured tailwind version.

This should be a backwards-compatible (aka non-breaking) change.

RobinBoers committed:
Support configuring version per profile

josevalim commented on lib/tailwind.ex:
```diff
@@ -163,8 +183,8 @@ defmodule Tailwind do
   """
-  def bin_version do
-    path = bin_path()
+  def bin_version(profile \\ :default) do
```
Are you required to have a default profile? I don't think so... so this will likely be a breaking change in practice.

RobinBoers replied:
If you do not have a default profile, this will now resolve to the path for the version configured 'globally' (aka `configured_version/0`), so that shouldn't be a breaking change.

The only "problem" is that after updating, people will have to rerun `mix tailwind.install`, because the path to the binary has changed to include the version string.

RobinBoers committed:
Rename back to `configured_version`

josevalim commented on lib/tailwind.ex:
```diff
@@ -105,18 +107,36 @@ defmodule Tailwind do
+  If not explicitly configured, falls back to `global_version/0`.
+  Raises if the given profile does not exist.
+  """
+  def configured_version!(profile) when is_atom(profile) do
+    :tailwind
+    |> Application.fetch_env!(profile)
+    |> Keyword.get(:version, global_version())
+  end
```
```suggestion
  def configured_version(profile) when is_atom(profile) do
    :tailwind
    |> Application.get_env(profile, [])
    |> Keyword.get(:version, global_version())
  end
```

RobinBoers committed:
Don't change default argument order

RobinBoers committed:
`configured_version` -> `configured_version!`

RobinBoers committed:
Collect & filter required versions before installing

josevalim commented on lib/tailwind.ex:
```diff
@@ -105,18 +107,36 @@ defmodule Tailwind do
+  @doc false
+  def profiles do
+    config_keys = [:version_check, :version, :target, :path]
+    :tailwind |> Application.get_all_env() |> Keyword.drop(config_keys)
```
And we should probably filter it to only return entries which have keyword lists as values.

```suggestion
    :tailwind |> Application.get_all_env() |> Keyword.drop(config_keys)
```

josevalim commented on lib/tailwind.ex:
```diff
@@ -146,8 +166,10 @@ defmodule Tailwind do
   """
-  def bin_path do
-    name = "tailwind-#{configured_target()}"
+  def bin_path, do: bin_path(configured_version())
+
+  def bin_path(version) do
```
```suggestion
  def bin_path(version \\ configured_version()) do
```

RobinBoers replied:
`configured_version/0` does an `Application.get_env` call. Does that work in default arguments?

josevalim replied:
The call is executed at runtime, as any other code, so it should be fine!

josevalim commented on lib/tailwind.ex:
```diff
@@ -228,11 +265,15 @@ defmodule Tailwind do
+    install(base_url, configured_version())
+  end
+
+  def install(base_url, version) do
```
We should either document it, doc false, or make it private.

josevalim commented on lib/tailwind.ex:
```diff
@@ -228,11 +265,15 @@ defmodule Tailwind do
 
   @doc """
-  Installs tailwind with `configured_version/0`.
+  Installs tailwind with `configured_version/1`.
```
```suggestion
  Installs tailwind with `configured_version/0`.
```

RobinBoers committed:
Update lib/tailwind.ex

Co-authored-by: José Valim <jose.valim@gmail.com>

RobinBoers committed:
Update lib/tailwind.ex

Co-authored-by: José Valim <jose.valim@gmail.com>

RobinBoers committed:
Filter profiles for keyword lists

RobinBoers committed:
Mark `@doc false` on `install/2`

RobinBoers committed:
Improve documentation

RobinBoers committed:
Fix broken test

RobinBoers committed:
Fix flaky test

The `--if-missing` argument causes this test to be flaky. Previously, when you had downloaded another version, the newer version would automatically be missing and thus always be installed.

Currently, that's no longer the case, because you can have multiple simultanous tailwind installs, so in this case (after the first test run), the newer version already exists.

RobinBoers commented:
Is there anything that needs to be fixed from my part to get this merged? Would be nice if we could switch back from our fork to the upstream for our projects.

RobinBoers committed:
Resolve merge conflicts

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -107,24 +112,46 @@ defmodule Tailwind do
+  Returns the configured tailwind version for a specific profile.
+
+  If not explicitly configured, falls back to `configured_version/0`.
+  Raises if the given profile does not exist.
```
It does not raise right now. If you want it to raise, you'd need to use `Application.fetch_env!` instead

RobinBoers replied:
I changed the implementation from raising to non-raising after [the suggestion from José](https://github.com/phoenixframework/tailwind/pull/123#discussion_r2060047976), but forgot to also update the docstring.

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -269,7 +319,7 @@ defmodule Tailwind do
   #  tailwindcss-macos-x64
   #  tailwindcss-windows-x64.exe
-  defp target do
+  defp system_target do
```
any reason to rename this?

RobinBoers replied:
I think my line of thought while reading the code was "hmmm what does this target function do, target seems vague, what kind of target does this function return?", then I read the function implementation and thought "ah it was the target of the current system, lets rename this to system_target to be more clear what it returns", without thinking about it much more.

But I agree that this is not my codebase and it is not my job/place to refactor or rename functions that are unrelated to the intent of the PR. Do you want me to change it back?

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -107,24 +112,46 @@ defmodule Tailwind do
 
+  @doc false
+  def profiles do
+    config_keys = [:version_check, :version, :target, :path]
```
we should probably move those to a more prominent location, e.g. `@config_keys` so it's more unlikely that we forget to add a key when we add a new option in the future. Overall, it's kind of suboptimal that the namespace is shared with profiles.

RobinBoers replied:
I agree. The reason we put profiles in this namespace and not in a separate 'profiles' key is to stay consistent with https://github.com/phoenixframework/esbuild, I think.

For now I moved the `config_keys` to a module attribute at the top of the module.

SteffenDE commented on lib/mix/tasks/tailwind.install.ex:
```diff
@@ -36,44 +40,69 @@ defmodule Mix.Tasks.Tailwind.Install do
-      case OptionParser.parse_head!(args, strict: valid_options) do
-        {opts, []} ->
-          {opts, Tailwind.default_base_url()}
+    if args |> try_install() |> was_successful?() do
```
I think there's an issue if a user only has a global version configured, but no profiles. In that case it would just do nothing.

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -21,6 +21,9 @@ defmodule Tailwind do
         ]
 
+  It is also possible to override the required tailwind CLI version on
+  profile-basis.
```
Should we add a note saying that per-profile versions are not supported if a `:path` is configured?

RobinBoers replied:
Yes, and in that case also emit a warning if a `:path` is configured with per-profile versions? Or we could support $version in the `:path`, like we also have in the download URL. (In that case also we still need to emit a warning when multiple profiles use different versions and no $version is found in the configured `:path` string).

SteffenDE replied:
I don't think we need to support placeholders in the path. Either we document and raise if a custom version is configured and path is set, or we require profiles with custom version to configure their own `:path`? It feels like a uncommon config though, so I'm fine with not handling it. Could also be added later.

SteffenDE commented on lib/mix/tasks/tailwind.install.ex:
```diff
@@ -36,44 +40,69 @@ defmodule Mix.Tasks.Tailwind.Install do
+          Mix.ensure_application!(:ssl)
+        end
+
+        Mix.Task.run("loadpaths")
```
This is probably fine, but it still feels wasteful to do those steps each time. You could first filter to only the versions that need installation and only then call ensure_application etc.

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -83,20 +86,22 @@ defmodule Tailwind do
 
-      configured_version = configured_version()
+      for {profile, config} <- profiles() do
+        configured_version = Keyword.get(config, :version, configured_version())
```
Since we have the new function now:

```suggestion
        configured_version = Tailwind.configured_version(profile)
```

RobinBoers replied:
Yes, but then we do an additional `Application.get_env` call. Seems wasteful if we already have the keyword list right there.

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -107,24 +112,46 @@ defmodule Tailwind do
   end
 
+  @doc """
+  Returns the configured tailwind version for a specific profile.
+
+  If not explicitly configured, falls back to `configured_version/0`.
+  Raises if the given profile does not exist.
+  """
+  def configured_version(profile) when is_atom(profile) do
+    :tailwind
+    |> Application.get_env(profile, [])
+    |> Keyword.get(:version, configured_version())
+  end
+
   @doc """
   Returns the configured tailwind target. By default, it is automatically detected.
   """
   def configured_target do
-    Application.get_env(:tailwind, :target, target())
+    Application.get_env(:tailwind, :target, system_target())
   end
```
The target detection needs to be version aware. Otherwise it could happen that a user configured a global v4, but a v3 in a profile and then system_target's ABI detection would pick the wrong v4 musl target suffix.

RobinBoers committed:
Fix wrong docstring

I changed the implementation from raising to non-raising after the suggestion from José, but forgot to also update the docstring.

RobinBoers committed:
Move `config_keys` to module attribute

RobinBoers committed:
Make targets version-aware

RobinBoers committed:
Refactor run function for performance

RobinBoers committed:
Fix `mix tailwind.install` not working if no profiles are configured

This now installs the given globally configured `version`, even if no profiles have been configured. This is not super useful because the `mix tailwind <profile>` command won't be usable, but maybe a valid use case when users only want to install the binary and call it themselves.

RobinBoers committed:
Fix typo

RobinBoers committed:
Add docstring to `bin_version/1`

RobinBoers committed:
Fix hardcoded path in warning

RobinBoers committed:
Make `File.chmod!` also a bang

RobinBoers committed:
Fix bug where computing `bin_path` on unsupported platforms crashes, even when `:path` is explicitly set

RobinBoers committed:
Make `configured_target/1` private

RobinBoers committed:
Refactor `configured_target/1` to be consistent with `configured_version/1`

RobinBoers committed:
Raise on boot when `:path` is set and one or more profiles configure `:version`

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -115,16 +133,68 @@ defmodule Tailwind do
+            do: profile
+
+      if profiles_with_version != [] do
+        raise """
```
```suggestion
        raise ArgumentError, """
```

SteffenDE commented on lib/tailwind.ex:
```diff
@@ -115,16 +133,68 @@ defmodule Tailwind do
+  end
+
+  defp target_for_version(version) when is_binary(version) do
+    Application.get_env(:tailwind, :target, system_target(version))
```
Right now, if someone overrides the target, it is set for all profiles, which might cause issues if one profile is Tailwind 4 and another one Tailwind 3. So similar to the global path. But if someone runs into this a separate PR to move path and target into a per-profile config would be welcome, so it can be overridden if necessary, falling back to the global or default.

SteffenDE committed:
Update lib/tailwind.ex

SteffenDE commented on test/tailwind_test.exs:
```diff
@@ -26,6 +26,30 @@ defmodule TailwindTest do
+      Application.delete_env(:tailwind, :pinned)
+    end)
+
+    assert_raise RuntimeError, ~r/cannot configure per-profile :version/, fn ->
```
```suggestion
    assert_raise ArgumentError, ~r/cannot configure per-profile :version/, fn ->
```

SteffenDE committed:
Apply suggestion from @SteffenDE
